### PR TITLE
Make sure subquery range is taken into account for selection

### PR DESCRIPTION
**Pull request by gouthamve**

### CODE REVIEW

**Comment by gouthamve** [comment]
This would make sure that `remote_read` and other downstream library users get the right `SelectParams` to select the data.

File: promql/engine.go [hunk 1]
```diff
@@ -552,6 +552,16 @@ func (ng *Engine) populateSeries(ctx context.Context, q storage.Queryable, s *Ev
 			Step:  durationToInt64Millis(s.Interval),
 		}
 
+		// We need to make sure we select the timerange selected by the subquery.
+		// TODO(gouthamve): cumulativeSubqueryOffset gives the sum of range and the offset
+		// we can optimise it by separating out the range and offsets, and substracting the offsets
+		// from end also.
+		subqOffset := ng.cumulativeSubqueryOffset(path)
+		if subqOffset > 0 {
```
Comment (brian-brazil) [inline promql/engine.go]: You don't need this if

**Comment by brian-brazil** [comment]
This should go against master.

**Comment by gouthamve** [comment]
@brian-brazil Addressed all comments.

File: promql/engine_test.go [hunk 2]
```diff
@@ -227,6 +227,167 @@ func TestQueryError(t *testing.T) {
 	}
 }
 
+// paramCheckerQuerier implements storage.Querier	which checks the start and end times
```
Comment (brian-brazil) [inline promql/engine_test.go]: stray tab here before which

File: promql/engine_test.go [hunk 3]
```diff
@@ -227,6 +227,167 @@ func TestQueryError(t *testing.T) {
...
+		query: "foo[2m] offset 2m",
+		start: 300,
+
+		paramStart: 60,
+		paramEnd:   180,
+	}, {
+		query: "foo[2m:1s]",
+		start: 300,
+
+		paramStart: 175, // 300 - 120 - 5(!)
```
Comment (brian-brazil) [inline promql/engine_test.go]: The -5 is correct here. We may need that data to calculate `foo` at the very start of the subquery.

File: promql/engine_test.go [hunk 4]
```diff
@@ -227,6 +227,167 @@ func TestQueryError(t *testing.T) {
...
+		paramEnd:   300,
+	}, {
+		query: "count_over_time((foo offset 10s)[2m:1s] offset 10s)",
+		start: 300,
+
+		paramStart: 155, // 300 - 120 - 5 - 10 - 10
+		paramEnd:   290,
+	},
+		// Range queries now.
+		{
```
Comment (brian-brazil) [inline promql/engine_test.go]: Indentation doesn't seem right here

**Comment by gouthamve** [comment]
Done :)

**Pull request merged**